src/parse/asp/exec.go

```diff
@@ -0,0 +1,165 @@
+package asp
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+)
+
+type execKey string
+type execValue interface {
+	// Found returns true when the value has been found in the cache
+	Found() bool
+}
+
+var (
+	// The output from exec() is memoized by default
+	memoizedCmdsLock sync.RWMutex
+	memoizedCmds     map[execKey]execValue
+)
+
+func init() {
+	memoizedCmdsLock.Lock()
+	defer memoizedCmdsLock.Unlock()
+
+	const initCacheSize = 16
+	memoizedCmds = make(map[execKey]execValue, initCacheSize)
+}
+
+// getCachedExecOutput returns the output if found, sets found to true if found,
+// and returns a held promise that must be either cancelled or completed.
+func getCachedExecOutput(args []string) (output string, found bool, promise execPromisePending) {
+	key := makeExecKey(args)
+
+	// We must perform manual locking for this first get test and can't rely on defer
+	memoizedCmdsLock.RLock()
+	rawVal, found := memoizedCmds[key]
+	if val, ok := rawVal.(*execData); found && ok {
+		// Simple case, the cmdArgs yielded a cachable output
+		memoizedCmdsLock.RUnlock()
+		return val.Value(), true, nil
+	}
+	memoizedCmdsLock.RUnlock()
+
+	// Promote to an exclusive, writer lock
+	memoizedCmdsLock.Lock()
+
+	// With an exclusive lock, attempt to see if the cache has been populated.
+	rawVal, found = memoizedCmds[key]
+	if !found {
+		// Create and return a new promise
+		promise := &execPromise{lock: &sync.Mutex{}, key: key}
+		promise.lock.Lock()
+		promise.cv = sync.NewCond(promise.lock)
+		memoizedCmds[key] = promise
+
+		memoizedCmdsLock.Unlock()
+		return "", false, promise
+	}
+
+	// Depending on the data found in the cache, return the value or block on the
+	// promise's CV.
+	switch val := rawVal.(type) {
+	case *execData:
+		// Simple case, the cmdArgs yielded a cachable output even though there was
+		// a race between dropping the lock above and now.
+		memoizedCmdsLock.Unlock()
+		return val.Value(), true, nil
+	case *execPromise:
+		promise := val
+		promise.cv.L.Lock()
+		for {
+			memoizedCmdsLock.Unlock()
+			promise.cv.Wait()
+			memoizedCmdsLock.Lock()
+			rawVal, found = memoizedCmds[key]
+			if !found {
+				// The promise lock that we're holding is no longer associated with a
+				// key in the memoization map (likely because the previous promise ran
+				// and exited unsuccessfully).  Create a promise and hope for the best
+				// on the second invocation..
+				promise := &execPromise{lock: &sync.Mutex{}, key: key}
+				promise.lock.Lock()
+				promise.cv = sync.NewCond(promise.lock)
+				memoizedCmds[key] = promise
+				return "", false, promise
+			}
+
+			if data, ok := rawVal.(*execData); ok {
+				promise.cv.L.Unlock()
+				memoizedCmdsLock.Unlock()
+				return data.Value(), true, nil
+			} else {
+				panic(fmt.Sprintf("unknown type in execValue: %T", val))
```

Comment:
This will only panic if there is a programmer error and we save something in the `memoizedCmds` map (e.g. we store an `int` or `struct{}` vs something that satisfies the type).  I'm happy to to move it to a `log.Fatalf()`, but have it as a `panic()` because it is a developer error, not a user error.

Reply:
I completely refactored this to be something significantly easier to understand: there is a map of cached values protected by a `sync.RWMutex` and a list of promises.  Cache misses get a promise that will return the resulting value from the concurrent command.  Future cache hits will just return the cached value itself.